quic/s2n-quic-core/src/crypto/tls/testing.rs

```diff
@@ -113,131 +126,198 @@ impl<S: tls::Session, C: tls::Session> Pair<S, C> {
-            1 => {
+    fn check_progress(&mut self) {
+        match (&self.client.context.state, &self.server.context.state) {
+            (ClientState::ClientHelloSent, ServerState::WaitingClientHello) => {
                 assert!(
-                    !self.server.1.initial.rx.is_empty(),
+                    !self.server.context.initial.rx.is_empty(),
                     "client should send ClientHello"
                 );
             }
-            2 => {
+            (ClientState::WaitingServerHello, ServerState::ClientHelloRead) => {
                 assert!(
-                    self.server.1.handshake.crypto.is_some(),
+                    self.server.context.handshake.crypto.is_some(),
                     "server should have handshake keys after sending the ServerHello"
                 );
-
                 assert!(
-                    self.server.1.application.crypto.is_some(),
+                    self.server.context.application.crypto.is_some(),
                     "server should have application keys after sending a ServerFinished"
                 );
-
-                assert!(!self.server.1.handshake_complete);
-                assert!(!self.client.1.handshake_complete);
+                assert!(!self.server.context.handshake_complete);
+                assert!(!self.client.context.handshake_complete);
             }
-            3 => {
+            (ClientState::ServerHelloRead, ServerState::WaitingClientFinish) => {
                 assert!(
-                    self.client.1.handshake.crypto.is_some(),
+                    self.client.context.handshake.crypto.is_some(),
                     "client should have handshake keys after reading the ServerHello"
                 );
                 assert!(
-                    self.client.1.application.crypto.is_some(),
+                    self.client.context.application.crypto.is_some(),
                     "client should have application keys after reading the ServerFinished"
                 );
                 assert!(
-                    self.client.1.handshake_complete,
+                    self.client.context.handshake_complete,
                     "client should complete the handshake"
                 );
             }
-            4 => {
+            (ClientState::ClientFinishSent, ServerState::ClientFinishRead) => {
                 assert!(
-                    self.server.1.handshake_complete,
+                    self.server.context.handshake_complete,
                     "server should finish after reading the ClientFinished"
                 );
+                // Transition to HandshakeComplete after this case
             }
-            _ => panic!("handshake made too many iterations"),
-        }
+            _ => unreachable!("handshake made too many iterations"),
+        };
+
+        self.client.context.state.transition();
+        self.server.context.state.transition();
```

Comment:
Would it be more clear if you move this outside `check_progress` and don't transition the states if the server hasn't finished processing the the async client hello callback?